packages/network-controller/src/NetworkController.test.ts

```diff
@@ -1,278 +1,4156 @@
+ * call to `getEIP1559Compatibility` that happens when the provider is set. This
+ * option is useful in tests that need a provider to get set but also call
+ * `getEIP1559Compatibility` on their own. In this case, since the
+ * `providerConfig` setter already calls `getEIP1559Compatibility` once, and
+ * since `getEIP1559Compatibility` is called out of band, the test may run with
+ * unexpected results. By stubbing `getEIP1559Compatibility` before setting the
+ * provider, the test is free to explicitly call it.
+ * @returns The set provider.
+ */
+async function setFakeProvider(
+  controller: NetworkController,
+  {
+    stubs = [],
+    stubLookupNetworkWhileSetting = false,
+    stubGetEIP1559CompatibilityWhileSetting = false,
+  }: {
+    stubs?: FakeProviderStub[];
+    stubLookupNetworkWhileSetting?: boolean;
+    stubGetEIP1559CompatibilityWhileSetting?: boolean;
+  } = {},
+): Promise<ProviderEngine> {
+  const fakeMetamaskProvider = buildFakeMetamaskProvider(stubs);
+  createMetamaskProviderMock.mockReturnValue(fakeMetamaskProvider);
+  const lookupNetworkMock = jest.spyOn(controller, 'lookupNetwork');
+  const lookupGetEIP1559CompatibilityMock = jest.spyOn(
+    controller,
+    'getEIP1559Compatibility',
+  );
+
+  if (stubLookupNetworkWhileSetting) {
+    lookupNetworkMock.mockResolvedValue(undefined);
+  }
+  if (stubGetEIP1559CompatibilityWhileSetting) {
+    lookupGetEIP1559CompatibilityMock.mockResolvedValue(undefined);
+  }
+
+  controller.providerConfig = buildProviderConfig();
+  await waitForResult(true, () => controller.provider !== undefined);
+  assert(controller.provider);
+
+  if (stubLookupNetworkWhileSetting) {
+    lookupNetworkMock.mockRestore();
+  }
+  if (stubGetEIP1559CompatibilityWhileSetting) {
+    lookupGetEIP1559CompatibilityMock.mockRestore();
+  }
+
+  return controller.provider;
+}
+
+/**
+ * Waits for a controller event to be emitted. If given a function, that
+ * function will be called first before waiting.
+ *
+ * @param messenger - The messenger suited for NetworkController.
+ * @param eventType - The type of event you want to wait for.
+ * @param rest - Either an options bag and a function, or a function, or
+ * nothing. The options bag may contain `count`, the number of events
+ * (default: 1); `filterEventPayload`, a function to filter the events; and
+ * `wait`, the amount of time to wait before assuming all events have occurred.
+ * The function will be called before the promise resolves.
+ * @returns A promise that resolves to the list of event payloads (optionally
+ * filtered) when the specified number of them have occurred.
+ */
+async function waitForPublishedEvents<E extends NetworkControllerEvents>(
+  messenger: NetworkControllerMessenger,
+  eventType: E['type'],
+  ...rest:
+    | [
+        {
+          count?: number;
+          filterEventPayload?: (payload: E['payload']) => boolean;
+          wait?: number;
+        },
+        () => void | Promise<void>,
```

Comment:
When reading this helper function in tests, it wasn't clear to me what this function was doing. Could we make this a named option, in the options bag? That would help explain the purpose of the function.

This would simplify this type signature and description as well.

Reply:
I guess the same goes for `waitForStateChanges` as well

Reply:
Okay, I've updated these functions per your feedback. Let me know if these changes make things more clear or if I need to go further: [26e6fcf](https://github.com/MetaMask/core/pull/1039/commits/26e6fcf4746d283e224e2fb9578b15cd7415d112)

Reply:
Thanks! It is a lot more readable, except that I found the name `after` quite confusing, because the function is called _before_ waiting for state changes :sweat_smile:

Reply:
Ah yes I can see how that would be confusing. Renamed here: [d94fa3e](https://github.com/MetaMask/core/pull/1039/commits/d94fa3e1cdd7c43a5cc7c3f09da34ef9ce5d0389)

Reply:
Oh nice, that is much more clear